### ISSUE TITLE
20% more gpu memory usage on resnet

### ISSUE THREAD

**Issue by csarofeen** [pytorch/pytorch#4081]
Been testing with resnet 50 when I noticed that a batch of 128 / gpu no longer fits in multi-gpu. I'm seeing around 20% more gpu memory being used now (see below).

'old version' commit hash https://github.com/pytorch/pytorch/commit/50009144c02155be6afd4570e93f453e73904a8e
'new version' commit hash https://github.com/pytorch/pytorch/commit/7ddcb91c7f84c3da8cc9f7fba28a3ae9ecd6cc45

```
resnet50    | batch 16/gpu | batch 32/gpu | batch 64/gpu
old version | 2.96 GiB     | 4.52 GiB     | 7.97 GiB
new version | 3.55 GiB     | 5.60 GiB     | 9.81 GiB
```

**Comment by ezyang** [comment]
I'm traveling this week and haven't had time to repro this, but despite your comment on the other issue, #3666 seems like the most high risk commit here. Do you think you might have a chance to bisect? If not, I will look at it next week.

**Comment by csarofeen** [comment]
result of git bisect:

```9cb8b437786648dfcf1e6a60dced70b6ee983311 is the first bad commit
commit 9cb8b437786648dfcf1e6a60dced70b6ee983311
Author: Sam Gross <colesbury@gmail.com>
Date:   Tue Nov 14 12:59:06 2017 -0500

    Split off in-place NN functions (#3683)
    
    For example, this splits threshold into threshold(), which is now
    never in-place, and threshold_() which is always in-place.
    
    This simplifies the in-place vs. non-in-place logic in
    gen_variable_type.py, which was bug-prone.

:040000 040000 41e00599aec0edd4ca83831b70b5bea6548f93e2 5dcb50f14bd811935e98e2286a1046944a80c015 M	aten
:040000 040000 7fe0482203d52e0739324180a4d7e2ea84be9e42 7b53a7090d7d0e567ec507ff45f544f260971482 M	docs
:040000 040000 1cce0769b77847d40e644bc69d34d690e05c411c de0d3f1b2fb41e3e9f7a0896e8f19038ff55b3a6 M	tools
:040000 040000 c7fc37d284a99153972a910558fbd95902fcb5bf 65483bdccdeb25bb9cac709ba5adaa8b4759e8d2 M	torch
```

@colesbury ^^ any ideas?

**Comment by colesbury** [comment]
Thanks for bisecting it. I'll take a look.

**Comment by colesbury** [comment]
OK, so the problem is that the auto-generated derivatives for the in-place functions are constructed such that they use the input (which has to be cloned, so it defeats the memory savings). They should be using the output instead.

I'll fix `gen_variable_type.py`.

@lantiga, when you implement #3874 you'll want to make sure that the derivative formulas for in-place functions use output/result instead of self. For example,

https://github.com/lantiga/pytorch/commit/eb14fa6d2fe578bb91f461516ddee163742a3145#diff-8026a1a389bfcf92e484ae272675c336R702

should be

```
self: threshold_backward(grad, output, threshold, value)
```

**Comment by lantiga** [comment]
Ok, so
```
self: elu_backward(grad, alpha, output)
```
is correct, while these change:
```
< self: hardtanh_backward(grad, self, min_val, max_val)
> self: hardtanh_backward(grad, output, min_val, max_val)

< self: leaky_relu_backward(grad, self, negative_slope)
> self: leaky_relu_backward(grad, output, negative_slope)

< self: rrelu_backward(grad, self, lower, upper, training, noise)
> self: rrelu_backward(grad, output, lower, upper, training, noise)

< self: threshold_backward(grad, self, threshold, value)
> self: threshold_backward(grad, output, threshold, value)
```

**Comment by colesbury** [comment]
@lantiga yeah, that looks right. Just make sure `test_nn.py` still passes.

**Comment by ngimel** [comment]
There are still memcopies for inplace relu's, even though #4096 is merged.

**Comment by apaszke** [comment]
@ngimel is this in forward or in backward?

**Comment by ngimel** [comment]
Memcopies are in forward, and generated code for backward still uses "self" rather than "output", I guess changes to ```gen_variable_type.py``` are necessary, as @colesbury was saying.
```
Tensor VariableType::threshold_backward(const Tensor & grad_output, const Tensor & self, Scalar threshold, Scalar value) const {
    profiler::RecordFunction profiler("threshold_backward");
    auto& grad_output_ = unpack(grad_output, "grad_output", 0);
    auto& self_ = unpack(self, "self", 1);
    std::shared_ptr<ThresholdBackwardBackward> grad_fn;
    auto flags = compute_flags({ grad_output, self });
    if (flags.requires_grad) {
      grad_fn = std::make_shared<ThresholdBackwardBackward>();
      grad_fn->next_functions = compute_next_functions({ grad_output, self });
      grad_fn->self_ = SavedVariable(self, false);
      grad_fn->threshold = threshold;
      grad_fn->value = value;
    }
    auto ret = as_variable(baseType->threshold_backward(grad_output_, self_, threshold, value));
    set_flags(ret, flags, grad_fn);
    if (jit::tracer::isTracing({ grad_output, self })) {
      jit::Node *n = jit::tracer::recordTrace( "threshold_backward", { grad_output, self }, { ret } );
      setattr(n, jit::stringToSymbol("threshold"), threshold);
      setattr(n, jit::stringToSymbol("value"), value);
    }
    return Tensor(std::move(ret));
}
Tensor VariableType::threshold_(Tensor & self, Scalar threshold, Scalar value) const {
    profiler::RecordFunction profiler("threshold_");
    auto& self_ = unpack(self, "self", 0);
    check_inplace(self);
    std::shared_ptr<ThresholdBackward> grad_fn;
    auto flags = compute_flags({ self });
    if (flags.requires_grad) {
      grad_fn = std::make_shared<ThresholdBackward>();
      grad_fn->next_functions = compute_next_functions({ self });
      grad_fn->self_ = SavedVariable(self.clone(), false);
      grad_fn->threshold = threshold;
      grad_fn->value = value;
    }
    baseType->threshold_forward_(self_, threshold, value);
    increment_version(self);
    set_flags(static_cast<Variable&>(self), flags, grad_fn, true);
    if (jit::tracer::isTracing({ self })) {
      jit::Node *n = jit::tracer::recordTrace( "threshold", { self }, { self } );
      setattr(n, jit::stringToSymbol("threshold"), threshold);
      setattr(n, jit::stringToSymbol("value"), value);
    }
    return self;
}
```

**Comment by lantiga** [comment]
We should use separate function names for non-inplace and inplace backwards if we want `output` just for `inplace`, like
```
- name: elu(Tensor self, Scalar alpha)
  self: elu_backward(grad, alpha, self)

- name: elu_(Tensor self, Scalar alpha)
  self: elu__backward(grad, alpha, output)
```

When I do that, I get `elu` tests failing (the other inplace functions seem to work):
```
ERROR: test_elu_inplace_gradgrad (__main__.TestNN)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_nn.py", line 3030, in test_elu_inplace_gradgrad
    gradcheck(func, [v])
  File "/Users/lantiga/miniconda2/lib/python2.7/site-packages/torch/autograd/gradcheck.py", line 182, in gradcheck
    return fail_test('for output no. %d,\n numerical:%s\nanalytical:%s\n' % (j, numerical, analytical))
  File "/Users/lantiga/miniconda2/lib/python2.7/site-packages/torch/autograd/gradcheck.py", line 167, in fail_test
    raise RuntimeError(msg)
RuntimeError: for output no. 0,
 numerical:[
 0.8945  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
 0.0000  1.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
 0.0000  0.0000  0.6910  0.0000  0.0000  0.0000  0.0000  0.0000
 0.0000  0.0000  0.0000  0.7863  0.0000  0.0000  0.0000  0.0000
 0.0000  0.0000  0.0000  0.0000  0.3021  0.0000  0.0000  0.0000
 0.0000  0.0000  0.0000  0.0000  0.0000  1.0000  0.0000  0.0000
 0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.3782  0.0000
 0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.4700
[torch.DoubleTensor of size 8x8]
]
analytical:(
 0.8885  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
 0.0000  1.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000
 0.0000  0.0000  0.6304  0.0000  0.0000  0.0000  0.0000  0.0000
 0.0000  0.0000  0.0000  0.7596  0.0000  0.0000  0.0000  0.0000
-0.0000  0.0000  0.0000  0.0000 -0.1969  0.0000  0.0000  0.0000
 0.0000  0.0000  0.0000  0.0000  0.0000  1.0000  0.0000  0.0000
 0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0276  0.0000
 0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.0000  0.2450
[torch.DoubleTensor of size 8x8]
,)


======================================================================
ERROR: test_elu_inplace_view (__main__.TestNN)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_nn.py", line 2929, in test_elu_inplace_view
    gradcheck(func, [v])
  File "/Users/lantiga/miniconda2/lib/python2.7/site-packages/torch/autograd/gradcheck.py", line 182, in gradcheck
    return fail_test('for output no. %d,\n numerical:%s\nanalytical:%s\n' % (j, numerical, analytical))
  File "/Users/lantiga/miniconda2/lib/python2.7/site-packages/torch/autograd/gradcheck.py", line 167, in fail_test
    raise RuntimeError(msg)
RuntimeError: for output no. 0,
 numerical:[
 1.0000  0.0000  0.0000  0.0000
 0.0000  0.3679  0.0000  0.0000
 0.0000  0.0000  1.0000  0.0000
 0.0000  0.0000  0.0000  1.0000
[torch.DoubleTensor of size 4x4]
]
analytical:(
 1  0  0  0
 0  0  0  0
 0  0  1  0
 0  0  0  1
[torch.DoubleTensor of size 4x4]
,)


======================================================================
FAIL: test_ELU (__main__.TestNN)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_nn.py", line 3904, in <lambda>
    setattr(TestNN, test_name, lambda self, test=test: test(self))
  File "/Users/lantiga/Desktops/Default/pytorch/test/common_nn.py", line 766, in __call__
    self._do_test(test_case, module, input)
  File "test_nn.py", line 101, in _do_test
    test_case.check_jacobian(module, input, self.jacobian_input)
  File "/Users/lantiga/Desktops/Default/pytorch/test/common_nn.py", line 630, in check_jacobian
    PRECISION
AssertionError: 0.3062127903660099 not less than or equal to 1e-05

----------------------------------------------------------------------
Ran 642 tests in 133.162s
```

**Comment by lantiga** [comment]
If we want the `backward` to get `output` as the argument just for `inplace` functions, shouldn't we give the inplace backward function a different name, like `threshold__backward` (two underscores)?

**Comment by ngimel** [comment]
I don't see problems with all (not just inplace) backward variants getting ```output``` as argument, but may be I'm missing something?

**Comment by apaszke** [comment]
Doesn't the backward of elu/sigmoid/... actually run a lot faster if you have the output (the derivatives are simple to express in terms of outputs)?

**Comment by ngimel** [comment]
Exactly, sigmoid/tanh require outputs for derivative computation (and yaml actually lists that). Generate sigmoid_ does not clone input, and saves output:
```
Tensor & VariableType::sigmoid_(Tensor & self) const {
    profiler::RecordFunction profiler("sigmoid_");
    auto& self_ = unpack(self, "self", 0);
    check_inplace(self);
    std::shared_ptr<SigmoidBackward> grad_fn;
    auto flags = compute_flags({ self });
    if (flags.requires_grad) {
      grad_fn = std::make_shared<SigmoidBackward>();
      grad_fn->next_functions = compute_next_functions({ self });
    }
    baseType->sigmoid_(self_);
    increment_version(self);
    set_flags(static_cast<Variable&>(self), flags, grad_fn, true);
    if (jit::tracer::isTracing({ self })) {
      jit::Node *n = jit::tracer::recordTrace( "sigmoid", { self }, { self } );
      (void)n;
    }
    if (grad_fn) {
      grad_fn->result_ = SavedVariable(self, true);
    }
    return self;
}
```

**Comment by lantiga** [comment]
I think I've got a fix for this.

**Comment by ngimel** [comment]
I think all that's needed is replace ```self``` with ```output``` here https://github.com/pytorch/pytorch/blob/8199edf5c1fb7073694e18954caa87602fbcce64/tools/autograd/derivatives.yaml#L734 and in similar places. Testing it now.

**Comment by lantiga** [comment]
I think there's a change in `gen_variable_type.py` for inplace to work properly, i.e. independently from the non-inplace derivative definition (you should be able to use `self` for the non inplace and `output` for the inplace):
```
--- a/tools/autograd/gen_variable_type.py
+++ b/tools/autograd/gen_variable_type.py
@@ -414,12 +414,7 @@ def load_derivatives(path, declarations_by_signature, declarations_by_name):
 
         declaration = declarations_by_name[defn_name][0]
         base_name = defn_name if not declaration['inplace'] else defn_name[:-1]
-        fwd_name = base_name + '_forward'
-
-        if declaration['inplace']:
-            declaration['base_name'] = fwd_name + '_'
-            declaration['derivative'] = declarations_by_name[base_name][0]['derivative']
-            return None
+        fwd_name = base_name + ('_forward' if not declaration['inplace'] else '_forward_')
 
         assert len(declarations_by_name[fwd_name]) == 1
```

**Comment by ngimel** [comment]
But a single _backward function is generated regardless of there being 2 declarations in the yaml file? I don't know how generation works.

**Comment by lantiga** [comment]
So, the _backward you see is actually the "forward" of the double backward. The actual backwards for, say, `threshold` and `threshold_` are set up when creating and configuring `grad_fn` inside the forward function.
Tests pass for me after the change above (will submit a PR in a minute): `clone` is gone from forward and `output` is saved after the call for all inplace functions, so we should be good.
Double backwards are out of scope I believe.

**Comment by ezyang** [comment]
Are all the fixes here merged to master now?

**Comment by ngimel** [comment]
#4184 is merged, #4199 is superceded by #4395, so this issue can be closed, and #4199 can be closed too? @lantiga?

**Comment by lantiga** [comment]
@ngimel yep, correct. Thank you for following up.
